### ISSUE TITLE
Docs: role dependency doesn't work as expected

### ISSUE THREAD

**Issue by liuyong2000** [ansible/ansible#17661]
##### ISSUE TYPE

- Bug Report
##### COMPONENT NAME

ansible/lib/ansible/playbook/role/**init**.py
ansible/lib/ansible/executor/play_iterator.py
##### ANSIBLE VERSION

```
ansible 2.1.0.0

```
##### CONFIGURATION

No change
##### OS / ENVIRONMENT

N/A
##### SUMMARY

In official ansible document site,
http://docs.ansible.com/ansible/playbooks_roles.html
It said,
Roles dependencies are always executed before the role that includes them, and are recursive. By default, roles can also only be added as a dependency once - if another role also lists it as a dependency it will not be run again.
but sometimes, it doesn't work as expected, the same role were executed multiple times. Maybe we need consider what role object is? the same role definition with different variable input may be considered different role object, but if we add conditional dependency for a role, that is we use "when", is the role with "when" different with role without "when"?
##### STEPS TO REPRODUCE

Suppose we have a playbook, which has a role named csf that depends on csfA/csfB/csfC, and csfB and csfA depends on csfC in some condition,

```
# main playbook
#ls.yml

---
- name: play1
  hosts: SSHproxy
  remote_user: root
  gather_facts: false
  roles:
    - {role: csf, taskN: task1 }

#csf role definition
#roles/csf/meta/main.yml

---
dependencies:
- role: csfC
- role: csfA
- role: csfB

# csfA role defintion
#roles/csfA/meta/main.yml

---
dependencies:
- {role: csfC, when: taskN=='task1'}

#roles/csfA/tasks/main.yml
- shell: ls


# csfB role definition
#roles/csfB/meta/main.yml

---
dependencies:
- {role: csfC, when: taskN=='task1'}

#roles/csfB/tasks/main.yml
- shell: ls

```

##### EXPECTED RESULTS

```
<!--- What did you expect to happen when running the steps above? -->
role csfC will be executed just one time

PLAY [play1] *******************************************************************

TASK [csfC : command] **********************************************************
changed: [SSHproxy]

TASK [csfA : command] **********************************************************
changed: [SSHproxy]

TASK [csfB : command] **********************************************************
changed: [SSHproxy]

PLAY RECAP *********************************************************************
SSHproxy : ok=3 changed=3 unreachable=0 failed=0
```
##### ACTUAL RESULTS

```
role csfC were executed two times.

PLAY [play1] *******************************************************************

TASK [csfC : command] **********************************************************
changed: [SSHproxy]

TASK [csfC : command] **********************************************************
changed: [SSHproxy]

TASK [csfA : command] **********************************************************
changed: [SSHproxy]

TASK [csfB : command] **********************************************************
changed: [SSHproxy]

PLAY RECAP *********************************************************************
SSHproxy : ok=4 changed=4 unreachable=0 failed=0
```

**Comment by liuyong2000** [comment]
I gave a fix to this problem, and passed my testing. My point is,  if a role running with same input parameters, they are considered same role object,  but "when" and "tags" doesn't belong to these parameter/variables.  If a role(role+parameter/variable input) has been executed, this role with "when" or different "when" should not be executed again, because they represent same role.
below is my fix for your reference.

```
diff -uNr ansible-2.1.0.0/lib/ansible/executor/play_iterator.py ansible-2.1.0.0-new/lib/ansible/executor/play_iterator.py
--- ansible-2.1.0.0/lib/ansible/executor/play_iterator.py       2016-05-25 21:00:55.000000000 +0800
+++ ansible-2.1.0.0-new/lib/ansible/executor/play_iterator.py   2016-09-20 15:39:55.655282489 +0800
@@ -243,7 +243,10 @@
         if task and task._role:
             # if we had a current role, mark that role as completed
             if s.cur_role and _roles_are_different(task._role, s.cur_role) and host.name in s.cur_role._had_task_run and not peek:
-                s.cur_role._completed[host.name] = True
+                if s.cur_role._ref_role:
+                   s.cur_role._ref_role._completed[host.name] = True
+                else:
+                   s.cur_role._completed[host.name] = True
             s.cur_role = task._role
             s.cur_dep_chain = task._block.get_dep_chain()

diff -uNr ansible-2.1.0.0/lib/ansible/playbook/role/__init__.py ansible-2.1.0.0-new/lib/ansible/playbook/role/__init__.py
--- ansible-2.1.0.0/lib/ansible/playbook/role/__init__.py       2016-05-25 21:00:55.000000000 +0800
+++ ansible-2.1.0.0-new/lib/ansible/playbook/role/__init__.py   2016-09-20 15:43:35.557282489 +0800
@@ -65,6 +65,7 @@

     _delegate_to = FieldAttribute(isa='string')
     _delegate_facts = FieldAttribute(isa='bool', default=False)
+    _ref_role = None

     def __init__(self, play=None):
         self._role_name        = None
@@ -82,6 +83,7 @@
         self._role_vars        = dict()
         self._had_task_run     = dict()
         self._completed        = dict()
+        self._ref_role = None

         super(Role, self).__init__()

@@ -100,17 +102,22 @@
             # We use frozenset to make the dictionary hashable.

             params = role_include.get_role_params()
+            params2 = params.copy()
             if role_include.when is not None:
                 params['when'] = role_include.when
             if role_include.tags is not None:
                 params['tags'] = role_include.tags
             hashed_params = hash_params(params)
+            hashed_params2 = hash_params(params2)
+            base_role = None
             if role_include.role in play.ROLE_CACHE:
                 for (entry, role_obj) in iteritems(play.ROLE_CACHE[role_include.role]):
                     if hashed_params == entry:
                         if parent_role:
                             role_obj.add_parent(parent_role)
                         return role_obj
+                    if hashed_params2 == entry:
+                       base_role = role_obj

             r = Role(play=play)
             r._load_role_data(role_include, parent_role=parent_role)
@@ -129,6 +136,10 @@
                     r.tags = new_tags

             play.ROLE_CACHE[role_include.role][hashed_params] = r
+            if base_role:
+               r._ref_role = base_role
+            else:
+               play.ROLE_CACHE[role_include.role][hashed_params2] = r
             return r

         except RuntimeError:
@@ -324,8 +335,14 @@
         Returns true if this role has been iterated over completely and
         at least one task was run
         '''
-
-        return host.name in self._completed and not self._metadata.allow_duplicates
+        host_list = []
+        if self._ref_role:
+           host_list = self._ref_role._completed
+        else:
+           host_list = self._completed
+
+        #return host.name in self._completed and not self._metadata.allow_duplicates
+        return host.name in host_list and not self._metadata.allow_duplicates

     def compile(self, play, dep_chain=None):

```

**Comment by gundalow** [comment]
Docs Issue:

* allow-duplicates

**Comment by acozine** [comment]
Addressed by #66371, published as https://docs.ansible.com/ansible/devel/user_guide/playbooks_reuse_roles.html#running-a-role-multiple-times-in-one-playbook.
